### PR TITLE
Gradually increasing HSTS max-age

### PR DESCRIPTION
**Pull request by joohoi**
This PR adds the functionality to enhance Apache configuration to include HTTP Strict Transport Security header with a low initial `max-age` value.

The `max-age` value will get increased on every (scheduled) run of `certbot renew` regardless of the certificate actually getting renewed, if the last increase took place longer than ten hours ago. The increase steps are visible in `constants.AUTOHSTS_STEPS`.

Upon the first actual renewal after reaching the maximum increase step, the `max-age` value will be made "permanent" and will get value of one year.

To achieve accurate `VirtualHost` discovery on subsequent runs, a comment with unique id string will be added to each enhanced `VirtualHost`.

### CODE REVIEW

**Comment by joohoi** [comment]
The tests will fail until #5913 is merged and this branch is brought up to date.

File: certbot-apache/certbot_apache/constants.py [hunk 1]
```diff
@@ -48,3 +48,16 @@
...
 HEADER_ARGS = {"Strict-Transport-Security": HSTS_ARGS,
                "Upgrade-Insecure-Requests": UIR_ARGS}
+
+AUTOHSTS_STEPS = [60, 300, 900, 3600, 21600, 43200, 86400]
+"""AutoHSTS increase steps: 5min, 15min, 1h, 6h, 12h, 24h"""
+
+AUTOHSTS_PERMANENT = 31536000
+"""Value for the last max-age of HSTS"""
+
+AUTOHSTS_FREQ = 36000
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/constants.py]: I agree. Would you think that 24h would suffice? Or should we increase the value even slower than that? I believe most of the issues with web applications should reveal themselves rather quickly, but we definitely need the increases to happen on a broader timespan.

File: certbot-apache/certbot_apache/configurator.py [hunk 2]
```diff
@@ -1529,6 +1587,152 @@ def enhance(self, domain, enhancement, options=None):
...
+
+    def _autohsts_update(self):
+        """
+        Increase the AutoHSTS values of VirtualHosts
+        """
+        try:
+            managed = self.storage.fetch("autohsts")
+        except KeyError:
+            # AutoHSTS not enabled
+            return
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/configurator.py]: That sounds like a good way to address this. Let's look into it as a part of the larger discussion about the enhancement interfaces.

File: certbot-apache/certbot_apache/configurator.py [hunk 3]
```diff
@@ -1529,6 +1587,152 @@ def enhance(self, domain, enhancement, options=None):
...
+        uniq_id = self.add_vhost_id(ssl_vhost)
+        self.save_notes += "Adding unique ID {0} to VirtualHost in {1}\n".format(
+            uniq_id, ssl_vhost.filep)
+        # Add the actual HSTS header
+        self.parser.add_dir(ssl_vhost.path, "Header", hsts_header)
+        note_msg = ("Adding gradually increasing HSTS header with initial value "
+                    "of {0} to VirtualHost in {1}\n".format(
+                        initial_maxage, ssl_vhost.filep))
+        self.save_notes += note_msg
+        self.save(note_msg)
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/configurator.py]: This will get addressed as a part of the enhancement interfaces refactor.

File: certbot/cli.py [hunk 4]
```diff
@@ -1104,6 +1104,12 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):  # pylint: dis
         default=flag_default("redirect"),
         help="Do not automatically redirect all HTTP traffic to HTTPS for the newly "
              "authenticated vhost. (default: Ask)")
+    helpful.add(
+        ["security", "enhance"],
+        "--auto-hsts", action="store_true", dest="auto_hsts",
+        default=flag_default("auto_hsts"),
+        help="Automatically increase Strict-Transport-Security header maxAge "
+             "value over time. Implies --hsts.")
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/cli.py]: Valid question. This isn't currently handled too well. I believe we should ignore `--hsts` and go with the `--auto-hsts`, as that's what the user most likely wants.

File: certbot-apache/certbot_apache/configurator.py [hunk 5]
```diff
@@ -2156,3 +2360,21 @@ def install_ssl_options_conf(self, options_ssl, options_ssl_digest):
...
         return common.install_version_controlled_file(options_ssl, options_ssl_digest,
             self.constant("MOD_SSL_CONF_SRC"), constants.ALL_SSL_OPTIONS_HASHES)
+
+    def generic_updates(self, _domain):
+        """
+        Interface method from interfaces.GenericUpdater to perform actions
+        when running certbot with "renew" verb, regardless of the lineage
+        being renewed.
+        """
+        self._autohsts_update()
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/configurator.py]: I was thinking of doing this later, as a part of a redesign of the complex enhancements but looks like it would make most sense to implement it as a part of this PR as well. I'm answering this in more detail in the comment about the interfaces, and old enhancement system.

File: certbot/main.py [hunk 6]
```diff
@@ -875,7 +875,7 @@ def enhance(config, plugins):
     :rtype: None
 
     """
-    supported_enhancements = ["hsts", "redirect", "uir", "staple"]
+    supported_enhancements = ["auto_hsts", "hsts", "redirect", "uir"]
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/main.py]: I was planning to do this later in a enhancements - redesign PR, but it looks like it wouldn't make much sense to use the old approach here if we're planning to refactor it soon anyway. 

Using new interfaces here sounds reasonable and like a good approach which we discussed earlier. I'm going to write up a few sketches as a base for the initial discussion, which I think we should use another forum for, as keeping track of discussions in PR review comments in GitHub is rather painful :)

In short, I think the updates in renewal time should keep getting triggered from the `interfaces.GenericUpdater.generic_updates()` even if calling the enhancement interface methods to keep the changes to Certbot core code minimal when adding new enhancement types as the call chain used for renewals is complicated enough as it is.

[the remark this answers is not in the document]
Comment (bmw) [reply certbot/main.py]: > I was planning to do this later in a enhancements - redesign PR, but it looks like it wouldn't make much sense to use the old approach here if we're planning to refactor it soon anyway.

Would you prefer we don't do it soon?

> In short, I think the updates in renewal time should keep getting triggered from the interfaces.GenericUpdater.generic_updates() even if calling the enhancement interface methods to keep the changes to Certbot core code minimal when adding new enhancement types as the call chain used for renewals is complicated enough as it is.

While I acknowledge, appreciate, and agree with you that tracking discussion in GH PRs is painful, I can't stop myself from providing feedback here. I hope you don't mind too much :smile:

I actually prefer we have a separate updater method for each enhancement. First, I think this allows for better documentation because the arguments each method takes can be cleanly and traditionally documented in a function rather than trying to do something like we do with `IInstaller.enhance` where we pass a dict and try and document in another file the values in the dict for each enhancement. If you were thinking of only calling `generic_updates` once for all updates, I think having separate methods helps prevent `generic_updates` from taking too many arguments over time.

Second, by having a single function, I think it causes each plugin to implement the code calling and/or determining which updates to call. Look at Apache/Nginx's implementation of `enhance`. By putting this code in Certbot, plugins don't have to implement it, simplifying their implementation and resulting in less code.

I'm certainly open to not doing things this way, but I do have a pretty strong preference for an update method per enhancement.

File: certbot/main.py [hunk 7]
```diff
@@ -911,8 +912,11 @@ def enhance(config, plugins):
     if not config.chain_path:
         lineage = cert_manager.lineage_for_certname(config, config.certname)
         config.chain_path = lineage.chain_path
-    le_client = _init_le_client(config, authenticator=None, installer=installer)
-    le_client.enhance_config(domains, config.chain_path, ask_redirect=False)
+    if oldstyle_enh:
+        le_client = _init_le_client(config, authenticator=None, installer=installer)
+        le_client.enhance_config(domains, config.chain_path, ask_redirect=False)
+    if enhancements.is_supported(config):
+        enhancements.enable(lineage, domains, installer, config)
```
Comment (bmw) [inline certbot/main.py]: This enhancement can currently only be enabled with the `enhance` subcommand. I think in the long term we also want it to work for `run` and `install`.

If this was on purpose and you want to do it later, that's fine, otherwise, I think this should be moved to `le_client.enhance_config`.

File: certbot/cli.py [hunk 8]
```diff
@@ -1124,6 +1129,16 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):  # pylint: dis
         default=flag_default("redirect"),
         help="Do not automatically redirect all HTTP traffic to HTTPS for the newly "
              "authenticated vhost. (default: Ask)")
+    for enh in enhancements._INDEX:  # pylint: disable=protected-access
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/cli.py]: Yeah, this makes it cleaner, and is now implemented.

File: certbot/cli.py [hunk 9]
```diff
@@ -1124,6 +1129,16 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):  # pylint: dis
...
+        helpful.add(enh["cli_groups"], enh["cli_flag"], action=enh["cli_action"],
+                    dest=enh["cli_dest"], default=enh["cli_flag_default"],
+                    help=enh["cli_help"])
+    helpful.add(
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/cli.py]: Sorry, this is some leftover cruft from the earlier iteration without the new enhancement system. Removed.

File: certbot/plugins/enhancements.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+from acme.magic_typing import Dict, List, Any  # pylint: disable=unused-import, no-name-in-module
+
+
+def is_supported(config):
+    """Checks if one or more of the requested enhancements are supported by
+    the enhancement interfaces."""
+    for enh in _INDEX:
+        enh_requested = hasattr(config, enh["cli_dest"])
+        enh_enabled = bool(getattr(config, enh["cli_dest"]))
+        if enh_requested and enh_enabled:
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/plugins/enhancements.py]: I like this approach, the now implemented generator sure yields (pun intended) cleaner and more readable code.

File: certbot/plugins/enhancements.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
+"""New interface style Certbot enhancements"""
+import abc
+import six
+
+from certbot import errors
+
+from acme.magic_typing import Dict, List, Any  # pylint: disable=unused-import, no-name-in-module
+
+
+def is_supported(config):
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/plugins/enhancements.py]: Changed, and it's better this way, thanks!

File: certbot/plugins/enhancements.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+                msg = ("Requested enhancement {} not supported by selected "
+                       "installer").format(enh["name"])
+                raise errors.NotSupportedError(msg)
+            # Run the enable method
+            getattr(installer, enh["enable_function"])(lineage, domains)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class AutoHSTSEnhancement(object):
+    """Example enhancement interface class for AutoHSTS"""
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/plugins/enhancements.py]: Oh yeah. The pydocs here are now much more descriptive. I forgot to change those.

File: certbot/plugins/enhancements.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+# This is used to configure internal new style enhancements in Certbot. These
+# enhancement interfaces need to be defined in this file. Please do not modify
+# this list from plugin code.
+_INDEX = [
+    {
+        "name": "AutoHSTS",
+        "cli_help": "Gradually increasing max-age value for HTTP Strict Transport "+
+                    "Security security header",
+        "cli_flag": "--autohsts",
+        "cli_flag_default": None,
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/plugins/enhancements.py]: Yeah, a very concrete example why it's best to handle this in a single place.

File: certbot/plugins/enhancements.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        "cli_groups": ["security", "enhance"],
+        "cli_dest": "auto_hsts",
+        "cli_action": "store_true",
+        "class": AutoHSTSEnhancement,
+        "updater_function": "update_autohsts",
+        "deployer_function": "deploy_autohsts",
+        "enable_function": "enable_autohsts"
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/plugins/enhancements.py]: I feel that in terms of arguments, the `lineage` should be the most powerful one that the enhancement could and would need in addition of the configuration object that is used in the installer initialization of course. I have a feeling that to provide different meaningful arguments for the enhancements, we would probably need to add those to renewal configuration as well.

File: certbot-apache/local-oldest-requirements.txt [hunk 15]
```diff
@@ -1,2 +1,2 @@
 -e acme[dev]
-certbot[dev]==0.21.1
+-e .[dev]
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/local-oldest-requirements.txt]: It's now temporarily 0.24.0, but I will change it before merging.

File: certbot-apache/certbot_apache/configurator.py [hunk 16]
```diff
@@ -2158,3 +2294,143 @@ def install_ssl_options_conf(self, options_ssl, options_ssl_digest):
...
+                    vhost = self.find_vhost_by_id(id_str)
+                except errors.PluginError:
+                    msg = ("VirtualHost with id {} was not found, unable to "
+                           "make HSTS max-age permanent.").format(id_str)
+                    logger.warning(msg)
+                    self._autohsts.pop(id_str)
+                    continue
+                if self._autohsts_vhost_in_lineage(vhost, lineage):
+                    vhosts.append(vhost)
+                    self._autohsts.pop(id_str)
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/configurator.py]: Yeah, handling the modification later on in the code now.

File: certbot/main.py (hunk 7 above)
Comment (joohoi) [reply certbot/main.py]: I will add this functionality after #6097 lands.

File: certbot/tests/enhancement_test.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+from certbot import updater
+
+from certbot.plugins import disco
+from certbot.plugins import enhancements
+from certbot.plugins import null
+
+import certbot.tests.util as test_util
+
+
+class EnhancementTest(test_util.ConfigTestCase):
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/tests/enhancement_test.py]: I agree, and will move the few tests that call main functions to `main_test.py`. However I feel that they all do have a reason to exist and should be able to pick up unintentional effects of some of the enhancement changes.

File: certbot/plugins/enhancements.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot/plugins/enhancements.py]: We agreed offline to add *args and **kwargs to the interface methods to allow some leeway if we need to add arguments without breaking the interfaces.

File: certbot/plugins/enhancement_test.py [hunk 18]
```diff
@@ -0,0 +1,179 @@
+"""Tests for new style enhancements"""
```
Comment (bmw) [inline certbot/plugins/enhancement_test.py]: We still need to do https://github.com/certbot/certbot/pull/5912#discussion_r194743614.

File: certbot/main.py [hunk 19]
```diff
@@ -908,11 +909,15 @@ def enhance(config, plugins):
         if not domains:
             raise errors.Error("User cancelled the domain selection. No domains "
                                "defined, exiting.")
+
+    lineage = cert_manager.lineage_for_certname(config, config.certname)
```
Comment (bmw) [inline certbot/main.py]: Oh Python. Nice catch moving this outside of the `if not config.chain_path` branch to avoid an ugly crash down below when calling `enhancements.enable`.

File: certbot-apache/setup.py [hunk 20]
```diff
@@ -8,7 +8,7 @@
 # acme/certbot version.
 install_requires = [
     'acme>0.24.0',
-    'certbot>=0.21.1',
+    'certbot>0.24.0',
```
Comment (bmw) [inline certbot-apache/setup.py]: This should now be set to `certbot>0.25.1`

File: certbot-apache/certbot_apache/configurator.py [hunk 21]
```diff
@@ -2158,3 +2294,173 @@ def install_ssl_options_conf(self, options_ssl, options_ssl_digest):
...
+        uniq_id = self.add_vhost_id(ssl_vhost)
+        self.save_notes += "Adding unique ID {0} to VirtualHost in {1}\n".format(
+            uniq_id, ssl_vhost.filep)
+        # Add the actual HSTS header
+        self.parser.add_dir(ssl_vhost.path, "Header", hsts_header)
+        note_msg = ("Adding gradually increasing HSTS header with initial value "
+                    "of {0} to VirtualHost in {1}\n".format(
+                        initial_maxage, ssl_vhost.filep))
+        self.save_notes += note_msg
+        self.save(note_msg)
```
Comment (bmw) [inline certbot-apache/certbot_apache/configurator.py]: We need to call restart after this.

Also, while I missed this in my previous review, the way we're using `save` here is different from the other AutoHSTS methods and the old enhancement system. Here we're calling `save` after enabling HSTS for each domain, with each save creating a checkpoint, rather than creating one save at the end. I think the single save is nicer rather than creating up to 100 checkpoints (and doing 100 restarts) in a single run.

**Comment by joohoi** [comment]
These are now fixed, and the new style enhancements have been enabled for `install` and `run` verbs as well. There are a few tests I'd like to write for `run` and `install` verbs making sure that we:
 - Fail early if installer doesn't support requested enhancements
 - Don't try to proceed if we don't have `config.certname` (case with `install` verb when used with custom key & cert)
 - Enhancements get triggered

Functionality wise it should be ready for the next round though.

File: certbot/tests/util.py [hunk 22]
```diff
@@ -379,3 +380,26 @@ def hold_lock(cv, lock_path):  # pragma: no cover
     cv.notify()
     cv.wait()
     my_lock.release()
+
+
+class MockInstallerAutoHSTS(enhancements.AutoHSTSEnhancement):
```
Comment (bmw) [inline certbot/tests/util.py]: I'm not sure if we need this class. I think we can get by with setting the spec attribute of the [Mock class](https://docs.python.org/3/library/unittest.mock.html#the-mock-class) to `enhancements.AutoHSTSEnhancement` so it'll pass `isinstance` checks. Is there something else we get here?

File: certbot/plugins/enhancement_test.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+    def setUp(self):
+        super(EnhancementTest, self).setUp()
+        self.mockinstaller = test_util.MockInstallerAutoHSTS()
+
+    @mock.patch('certbot.plugins.selection.choose_configurator_plugins')
+    @test_util.patch_get_utility()
+    def test_enhancement_updates(self, _, mock_select):
+        mock_select.return_value = (self.mockinstaller, None)
+        updater.run_generic_updaters(self.config, mock.MagicMock(), None)
```
Comment (bmw) [inline certbot/plugins/enhancement_test.py]: Sorry to keep nitpicking this, but I think tests that are calling into other modules like `updater` should be deleted or moved to that module. I think we should just be testing the functions in `enhancements` here. The tests should be very simple.

File: certbot/plugins/enhancement_test.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
+"""Tests for new style enhancements"""
```
Comment (bmw) [inline certbot/plugins/enhancement_test.py]: We should rename this file to the plural "enhancements" to match the name of the module.

File: certbot/plugins/enhancement_test.py (hunk 23 above)
Comment (joohoi) [reply certbot/plugins/enhancement_test.py]: This makes a lot of sense, and the tests have now been refactored as such.

File: certbot/tests/util.py (hunk 22 above)
Comment (bmw) [reply certbot/tests/util.py]: But what's wrong with `installer.update_autohsts.call_count`, `installer.update_autohsts.called`, `installer.update_autohsts.call_args`, etc?

This seems more conventional and takes less code. The only change with the current approach I see is you access these values on `installer.update_counter` rather than `installer.update_autohsts`.

Comment (joohoi) [reply certbot/tests/util.py]: You have a point, and the tests are now refactored. The tests for `interfaces.GenericUpdater` and `interfaces.RenewDeployer` were similar and handled in the same files, so I went ahead and refactored those too for consistency.

File: certbot/tests/main_test.py [hunk 25]
```diff
@@ -1682,5 +1686,68 @@ def test_plugin_selection_error(self, _rec, mock_choose, mock_pick):
...
+            self._call, ['enhance', '--auto-hsts'])
+
+    def test_enhancement_enable_conflict(self):
+        self.assertRaises(
+            errors.Error,
+            self._call, ['enhance', '--auto-hsts', '--hsts'])
+
+    @mock.patch('certbot.main.plug_sel.record_chosen_plugins')
+    @mock.patch('certbot.main.plug_sel.pick_installer')
+    def test_install_enhancement_not_supported(self, mock_inst, _rec):
```
Comment (bmw) [inline certbot/tests/main_test.py]: One more test location nit. This docstring for this test class is "Tests for certbot.main.enhance.", but this test and the two below it aren't testing the enhance method. I think the one calling `run` should be moved to `RunTest`. We don't have an `InstallTest` class, but I'd prefer we created one and put these methods there.

File: certbot/tests/renewupdater_test.py [hunk 26]
```diff
@@ -75,6 +62,68 @@ def test_deployer_skip_dry_run(self, mock_log):
         self.assertEquals(mock_log.call_args[0][0],
                           "Skipping renewal deployer in dry-run mode.")
 
+    @mock.patch('certbot.plugins.selection.choose_configurator_plugins')
+    @test_util.patch_get_utility()
```
Comment (bmw) [inline certbot/tests/renewupdater_test.py]: I don't think we need to call `patch_get_utility` here or on any of the tests below. I just ran tests with them removed and they still passed.

File: certbot/tests/main_test.py [hunk 27]
```diff
@@ -1682,5 +1686,76 @@ def test_plugin_selection_error(self, _rec, mock_choose, mock_pick):
...
+            errors.NotSupportedError,
+            self._call, ['enhance', '--auto-hsts'])
+
+    def test_enhancement_enable_conflict(self):
+        self.assertRaises(
+            errors.Error,
+            self._call, ['enhance', '--auto-hsts', '--hsts'])
+
+    @mock.patch('certbot.main.plug_sel.choose_configurator_plugins')
+    def test_run_enhancement_not_supported(self, mock_choose):
```
Comment (bmw) [inline certbot/tests/main_test.py]: It looks like part of my comment at https://github.com/certbot/certbot/pull/5912#discussion_r196519729 was missed. Can we move this to the `RunTest` class in this file?

Comment (joohoi) [reply certbot/tests/main_test.py]: I'm very sorry, this is getting embarrassing. I'm apparently too eager to get the PR out of my hands. Moving the tests now.
